Porting STM32 Ethernet HAL driver

gleichdick opened the pull request:
This is a follow-up of #45260 to fix #41380.

A new Kconfig option (<del>`ETH_STM32_USE_HAL_DRIVER_WITH_NEW_API`</del>`ETH_STM32_HAL_API_V2`) is introduced to switch to the non-legacy ETH HAL driver. Furthermore, multiple receive buffers can be used for one single packet, which means that the rx buffer size can be reduced and doesn't have to match the maximum ethernet packet length anymore.

Future work:
 - [ ] Make rx buffer size selectable by config option (?)
 - [x] Support multiple buffers for TX path as well for the same reason
 - [ ] Use new PTP HAL

Depends on zephyrproject-rtos/hal_stm32#139

Commit:
drivers: ethernet: stm32: Add Kconfig to select new HAL API

STM32H7X and STM32F4X ETH HAL Drivers now provide a new api.
This commit only adds a new Kconfig option

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

Commit:
drivers: ethernet: stm32: Additional structures for new HAL driver

They allow managing the tx and rx buffers from both
driver parts (zephyr and stm32cube)

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

Commit:
drivers: ethernet: stm32: Add transmit callbacks

These callbacks are also used by the stm32cube HAL driver

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

Commit:
drivers: ethernet: stm32: initialization routine for the new driver

Setup has to be adapted, too

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

bbilas commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -152,4 +152,11 @@ config ETH_STM32_HAL_PTP_CLOCK_INIT_PRIO
 
+config ETH_STM32_USE_HAL_DRIVER_WITH_NEW_API
+	bool "Use new HAL driver"
+	default no
```
That is not needed, please remove it. Please see https://docs.zephyrproject.org/1.14.0/guides/kconfig/index.html#redundant-defaults

bbilas replied:
That is not needed, please remove it. Please see https://docs.zephyrproject.org/1.14.0/guides/kconfig/index.html#redundant-defaults

erwango commented:
@gleichdick Thanks for this updated version!
In order to ease the use of your PR, please add a commit which will update manifest to use the modified hal_stm32:
In `west.yml`, do:
```
    - name: hal_stm32
-      revision: 6253094a4147cbd30d2b466c975eb4013ac6be90
+     revision: pull/139/head
      path: modules/hal/stm32
      groups:
        - hal
```

erwango commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -152,4 +152,11 @@ config ETH_STM32_HAL_PTP_CLOCK_INIT_PRIO
 
 endif # SOC_SERIES_STM32F7X || SOC_SERIES_STM32H7X
 
+config ETH_STM32_USE_HAL_DRIVER_WITH_NEW_API
```
Also, as this is a somewhat important option for the short term future, please put it much higher in the file (eg line 17).
Also potentially there will be impact on the subsequent options, so better to declare it before them.

Commit:
drivers: ethernet: stm32: Add receive callbacks

Callbacks which will be used by stm32cube driver

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

gleichdick commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -439,7 +562,18 @@ static struct net_pkt *eth_rx(const struct device *dev, uint16_t *vlan_tag)
+		/* no frame available */
+		return NULL;
+	}
+
```
By looking in the disassembly, it seems that gcc 10.3.1 optimizes away the case that a packet has been received. I suspect that this is due to some undefined behaviour (UB). Compiling with `-fno-strict-aliasing` does not change anything, but with `-Og` it seems to be fine (as `eth_rx` is not inlined)

gleichdick replied:
Actually, this is not true. `eth_rx` is inlined but nothing is optimized away

gleichdick commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -439,7 +562,18 @@ static struct net_pkt *eth_rx(const struct device *dev, uint16_t *vlan_tag)
+		/* no frame available */
+		return NULL;
+	}
+
```
By looking in the disassembly, it seems that gcc 10.3.1 optimizes away the case that a packet has been received. I suspect that this is due to some undefined behaviour (UB). Compiling with `-fno-strict-aliasing` does not change anything, but with `-Og` it seems to be fine (as `eth_rx` is not inlined)

gleichdick reviewed (commented):
I'm finally able to run some tests on a Nucleo H743. The new driver seems to work in general (tested the telnet example), but suffers some initial packet losses <del>and does not work at all with `-O2` enabled</del>.
```
$  ping  stm32
PING stm32 (192.0.2.1) 56(84) bytes of data.
64 bytes from stm32 (192.0.2.1): icmp_seq=9 ttl=64 time=1.09 ms
64 bytes from stm32 (192.0.2.1): icmp_seq=10 ttl=64 time=0.577 ms
64 bytes from stm32 (192.0.2.1): icmp_seq=11 ttl=64 time=0.517 ms
64 bytes from stm32 (192.0.2.1): icmp_seq=12 ttl=64 time=0.452 ms
64 bytes from stm32 (192.0.2.1): icmp_seq=13 ttl=64 time=0.511 ms
```
First 8 ping packages are lost

gleichdick commented:
The package loss was due to the arp cache. No package losses when clearing the host arp cache after mcu reset.

erwango commented:
@gleichdick Is this ready for review ?

erwango commented:
@gleichdick Please let us know the status for this work. Thanks!

Commit:
drivers: ethernet: stm32: use modified hal_stm32 manifest

the hal package was also adapted

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

gleichdick commented:
Yes, I think this ready now. Thank you.

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -89,7 +89,27 @@ static ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB] __eth_stm32_desc;
+static struct eth_stm32_rx_buffer_header dma_rx_buffer_header[ETH_RXBUFNB];
+static bool dma_tx_buffer_used[ETH_TXBUFNB];
+
+#endif /* defined(CONFIG_ETH_STM32_HAL_API_V2) */
```
Nitpick:
```suggestion
#endif /* CONFIG_ETH_STM32_HAL_API_V2 */
```
Apply everywhere

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -270,41 +274,61 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
+		return -EIO;
+	}
+
+	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
```
Changes to existing code should be done in a dedcaited commit with justification. Ideally a dedicated PR if there are multiple.

gleichdick replied:
Here it is: 34898e22d8a7

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -284,18 +319,8 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
-
-buffer_found:
+	ctx.first_tx_buffer_index = allocate_tx_buffer();
+	buf_header = &dma_tx_buffer_header[ctx.first_tx_buffer_index];
```
Can you squash this initial commit, so we don't review code that is later deleted ?

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -191,40 +309,73 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
 
 	heth = &dev_data->heth;
 
-	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
-
 	total_len = net_pkt_get_len(pkt);
-	if (total_len > ETH_TX_BUF_SIZE) {
+	if (total_len > (ETH_TX_BUF_SIZE * ETH_TXBUFNB)) {
 		LOG_ERR("PKT too big");
-		res = -EIO;
-		goto error;
+		return -EIO;
 	}
 
+	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
```
Changes to existing code should be done in dedicated commit and explained.

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -89,7 +89,108 @@ static ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB] __eth_stm32_desc;
+};
+
+struct eth_stm32_tx_buffer_header {
+	ETH_BufferTypeDef hal_def;
```
`hal_def` is too generic, what about:
```suggestion
	ETH_BufferTypeDef tx_buff;
```

erwango reviewed (changes requested):
Here is a first round of comments.

Also, can you check about commits formating: https://static.sched.com/hosted_files/zephyr2022/a3/HOWTO_Get_Your_Zephyr_Patches_Merged_ZDS_2022.pdf ?

Commit:
drivers: ethernet: stm32: check tx packet size before locking mutex

The transmit mutex is not needed while checking the packet size,
so we acquire the lock after checking the size and
return early if it is too big.

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

Commit:
drivers: ethernet: stm32: Enable transmitting with new driver

Use the new API in the transmit method

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

Commit:
drivers: ethernet: stm32: Enable receiving with new driver

Use the new API in the receive method

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

Commit:
drivers: ethernet: stm32: Make tx semaphore available for new api

The new HAL API also has a TxCpltCallback, so we need the semaphore too.

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

Commit:
drivers: ethernet: stm32: Disable autonegation for new api

As it is not supported

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

gleichdick commented:
I also had an idea to add
```c
#ifdef CONFIG_ETH_STM32_HAL_API_V2
#error "This KConfig option is not fully implemented"
#endif
```
In the first commit and to remove it in the last. Even though the code compiles after each commit (at least it should), the new api only works after the last commit.

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -812,6 +812,10 @@ void HAL_ETH_TxCpltCallback(ETH_HandleTypeDef *heth_handle)
+#endif /* CONFIG_SOC_SERIES_STM32H7X || CONFIG_ETH_STM32_HAL_API_V2 */
+
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
```
Let's move this to the next commit for simplicity.

gleichdick replied:
Okay

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -199,12 +199,12 @@ static HAL_StatusTypeDef read_eth_phy_register(ETH_HandleTypeDef *heth,
 						uint32_t PHYReg,
 						uint32_t *RegVal)
 {
-#if defined(CONFIG_SOC_SERIES_STM32H7X)
+#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_ETH_STM32_HAL_API_V2)
 	return HAL_ETH_ReadPHYRegister(heth, PHYAddr, PHYReg, RegVal);
 #else
 	ARG_UNUSED(PHYAddr);
 	return HAL_ETH_ReadPHYRegister(heth, PHYReg, RegVal);
-#endif /* CONFIG_SOC_SERIES_STM32H7X */
+#endif /* CONFIG_SOC_SERIES_STM32H7X || CONFIG_ETH_STM32_HAL_API_V2 */
```
Link with the current commit ?

gleichdick replied:
Could you elaborate on that?

erwango replied:
Title of the commit is "drivers: ethernet: stm32: Add transmit callbacks"

gleichdick replied:
Thanks! Like this?

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -815,6 +815,31 @@ void HAL_ETH_TxCpltCallback(ETH_HandleTypeDef *heth_handle)
+{
+	__ASSERT_NO_MSG(heth != NULL);
+
+	const uint32_t error_code = HAL_ETH_GetError(heth);
```
Why `const`?

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -304,38 +310,72 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
 	heth = &dev_data->heth;
 
 	total_len = net_pkt_get_len(pkt);
-	if (total_len > ETH_TX_BUF_SIZE) {
+	if (total_len > (ETH_TX_BUF_SIZE * ETH_TXBUFNB)) {
```
Move this in a dedicated commit as it impacts existing code

gleichdick replied:
Done

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -153,6 +153,41 @@ void HAL_ETH_RxLinkCallback(void **pStart, void **pEnd, uint8_t *buff, uint16_t
+	}
+}
+
+/* allocate a tx buffer and mark it as used */
+static inline uint16_t allocate_tx_buffer(void)
+{
+	for (;;) {
+		for (uint16_t index = 0; index < ETH_TXBUFNB; index++) {
+			if (!dma_tx_buffer_header[index].used) {
+				dma_tx_buffer_header[index].used = true;
+				return index;
+			}
+		}
+		k_yield();
+	}
+}
```
Seem this function belongs to one of the next commits

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -304,38 +310,72 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
 		dma_tx_desc->DESC2 |= ETH_DMATXNDESCRF_TTSE;
-#else
+#else /* CONFIG_SOC_SERIES_STM32H7X */
```
The comment on else is confusing

gleichdick replied:
Removed

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -347,15 +387,22 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
 		goto error;
 	}
 
-#if defined(CONFIG_SOC_SERIES_STM32H7X)
```
Isn't this breaking existing NON H7 code ?

gleichdick replied:
Yes, it is. Good catch!

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -696,15 +722,36 @@ static struct net_pkt *eth_rx(const struct device *dev, uint16_t *vlan_tag)
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+	for (rx_header = (struct eth_stm32_rx_buffer_header *)appbuf;
+			rx_header; rx_header = rx_header->next) {
+		const size_t index = rx_header - &dma_rx_buffer_header[0];
```
why `const` ?

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -191,40 +309,73 @@ static int eth_tx(const struct device *dev, struct net_pkt *pkt)
+	k_mutex_lock(&dev_data->tx_mutex, K_FOREVER);
+
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+
```
Nitpick: looking to existing style inside this file, no spaces are needed around these lines.

erwango reviewed (changes requested):
Thanks for the update.

One additional comment regarding the commit messages.
I suggest to update titles in the following way:
`drivers: ethernet: stm32: bla`

See https://youtu.be/mnVNLM7ynMQ?t=936

Commit:
drivers: ethernet: stm32: Transmit frames larger than one buffer

Now, multiple tx buffers can be used to send packets,
so that the packet size can exceed tx buffer size.

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

Commit:
drivers: ethernet: stm32: Add Error Callback

The new HAL API provides a different error propagation method

Signed-off-by: Bjarne von Horn <B.von_horn@wzl.rwth-aachen.de>

GeorgeCGV commented:
> @GeorgeCGV As I recall you're an STM32 eth user. Would you ave time to review and test this ?

I am sorry, but I am not using STM32 ETH.  I am not sure where that idea comes from.

erwango commented:
> I am sorry, but I am not using STM32 ETH. I am not sure where that idea comes from.

Ok, my bad

erwango commented:
@gleichdick Would you mind rebasing this ?
I'd like to merge it soon. It would be disabled by default, so no risk, and easier to test.

erwango commented:
@gleichdick I'm getting errors when trying to build with new API:

```
$ west build  -b nucleo_h723zg samples/net/dhcpv4_client
```

```
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:95:14: error: 'ETH_RX_BUF_SIZE' undeclared here (not in a function); did you mean 'ETH_STM32_RX_BUF_SIZE'?
   95 | BUILD_ASSERT(ETH_RX_BUF_SIZE % 4 == 0, "Rx buffer size must be a multiple of 4");
      |              ^~~~~~~~~~~~~~~
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:95:14: error: expression in static assertion is not an integer
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c: In function 'HAL_ETH_RxLinkCallback':
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:139:9: error: type defaults to 'int' in type name [-Werror=implicit-int]
  139 |         size_t index = (RxBufferPtr)buff - &dma_rx_buffer[0];
      |         ^~~~~~
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:139:42: error: invalid operands to binary - (have 'int' and 'uint8_t (*)[1528]' {aka 'unsigned char (*)[1528]'})
  139 |         size_t index = (RxBufferPtr)buff - &dma_rx_buffer[0];
      |                        ~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~
      |                        |                   |
      |                        int                 uint8_t (*)[1528] {aka unsigned char (*)[1528]}
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c: In function 'eth_tx':
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:356:33: error: 'ETH_TX_BUF_SIZE' undeclared (first use in this function); did you mean 'ETH_STM32_TX_BUF_SIZE'?
  356 |         while (remaining_read > ETH_TX_BUF_SIZE) {
      |                                 ^~~~~~~~~~~~~~~
      |                                 ETH_STM32_TX_BUF_SIZE
/local/mcu/zephyrproject/zephyr/drivers/ethernet/eth_stm32_hal.c:356:33: note: each undeclared identifier is reported only once for each function it appears in

```

gleichdick commented:
cc9b6bb0b4e removed `ETH_{R,T}X_BUF_SIZE` definitions, I replaced them with `ETH_STM32_{R,T}X_BUF_SIZE`
